### PR TITLE
Transform mode for image layer

### PR DESCRIPTION
**Pull request by jojoelfe**
# Description
This builds on #3577 and starts enabling some of the functionality in the Qt-viewer. The main points are:
- This implements a "transform" mode for the image layer
- Keybindings are hooked up to allow zooming with "space" and fixed aspect ratio scaling with "shift"
- A lot of errors related to nD-images are fixed

transform

There are still quite a few rough edges, especially around nD-images, but I thought it might be a good point to get some feedback about what expected behavior with nD-images is (which kind of makes my head explode).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

- [x] This change requires a documentation update

# TODO

- [x] Remove button in qt_image_control
- [x] Vertex color-change on mouseover
- [x] Update box affine with layer affine
- [x] Preserve transform of not displayed dimensions
- [x] Keyboard shortcut to reset transform

# How has this been tested?

I think this passes most tests, for some reason some tests on my machine always fail. I think there also should be some tests for the new behaviors.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by sofroniewn** [comment]
This is so awesome! Just gave it a play with. Two minor things are
- it would be great if the corners of the box had a white fill and then changed to blue on hover. i believe we do that in our other layers (can come later if a pain)
- it would be nice to have a "reset" shortcut, maybe cmd+shift+R

I think we can probably get this in now for image as is. I think it would be nice to extend to labels asap (which is basically the same as image). We could extend to all layers too, but that can maybe come later

> There are still quite a few rough edges, especially around nD-images, but I thought it might be a good point to get some feedback about what expected behavior with nD-images is (which kind of makes my head explode).

I didn't notice any aside from after rotation - which is a known limitation. Everything seemed to make sense to me, maybe @alisterburt or @brisvag can check this out with some more real world data.

Code implementation looks very good too. I can take a finer look when it's ready, but I didn't see anything major off.

This will be so awesome for folks doing affine registration cc @jni @GenevieveBuckley!!

**Comment by jojoelfe** [comment]
> I think we can probably get this in now for image as is. I think it would be nice to extend to labels asap (which is basically the same as image). We could extend to all layers too, but that can maybe come later

Actually, I was wondering how people would feel about moving the mode property into the base layer (together with the Mode-enum). This way the same code could be reused for all layers.


>     * it would be nice to have a "reset" shortcut, maybe cmd+shift+R

Yes, a keyboard shortcut would be nice. I was thinking "Esc" or "Backspace", but not sure if they are already taken. I also thought it would be neat if right-clicking the transform-mode icon would show a context menu with an option to "Reset Transform". I have not looked into how to implement it.



>     * it would be great if the corners of the box had a white fill and then changed to blue on hover. i believe we do that in our other layers (can come later if a pain)

That would be nice as well as appropriate cursors. I think I saw a PR recently that implemented a custom cursor

**Comment by sofroniewn** [comment]
> Actually, I was wondering how people would feel about moving the mode property into the base layer (together with the Mode-enum). This way the same code could be reused for all layers.

Yes, I'd be fine with that. Thoughts @jni @tlambert03? I think it has been discussed elsewhere too. We can decide later if we want to expose the resize button on every layer controls in the GUI

> Yes, a keyboard shortcut would be nice. I was thinking "Esc" or "Backspace", but not sure if they are already taken.

Yeah, they are in some layers. I think I still prefer cmd-shift-r as I see this is a reset (like an undo), whereas I see esc/ backspace as stopping an action or exiting something

> I also thought it would be neat if right-clicking the transform-mode icon would show a context menu with an option to "Reset Transform". I have not looked into how to implement it.

Yeah we could do that, though I think that should be it's own PR. We don't expose any layer specific context menu items from the canvas yet I think. If we do this I think we might want to leverage some of the context machinery that @tlambert03 has been working on for right clicking in the layer list. We could actually add this to the layer list context menu right away. Still maybe best to do in a small followup PR, but that should be very easy

> That would be nice as well as appropriate cursors. I think I saw a PR recently that implemented a custom cursor

Yes agreed - perfect

**Comment by tlambert03** [comment]
> > Actually, I was wondering how people would feel about moving the mode property into the base layer (together with the Mode-enum). This way the same code could be reused for all layers.

> Yes, I'd be fine with that. Thoughts @jni @tlambert03? I think it has been discussed elsewhere too

I'd be ok with that.  I'm slightly less sure about the new mode buttons at the top of the image layer controls in the Gui.  Precious space...

**Comment by tlambert03** [comment]
> I also thought it would be neat if right-clicking the transform-mode icon would show a context menu with an option to "Reset Transform". I have not looked into how to implement it.

I think we're getting close to a place where a `Layer` menu (in the main menu at the top) would be appropriate.  (And also close(r) to a global command palette).  This seems very well suited to those things as well

**Comment by sofroniewn** [comment]
> I'd be ok with that. I'm slightly less sure about the new mode buttons at the top of the image layer controls in the Gui. Precious space...

Yeah, we could get this PR in without the button first, and then decide on button in the followup PR. I am a little nervous about that as well, but don't want to hold this up for a lengthy design debate

> I think we're getting close to a place where a Layer menu (in the main menu at the top) would be appropriate. 

Yeah that could be cool too

**Comment by jojoelfe** [comment]
> I'd be ok with that. I'm slightly less sure about the new mode buttons at the top of the image layer controls in the Gui. Precious space...

If the transform works consistently across all layers, how about putting the button for it into the layerlist:
tr_pos

**Comment by sofroniewn** [comment]
> If the transform works consistently across all layers, how about putting the button for it into the layerlist:

That looks cool! Once nice thing could then be to adjust the transforms simultaneously for all selected layers. That could be super helpful if aligning multiple channels. Curious what others think here too - but it makes sense to me!!

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Agreed that this is important. I think it needs to be addressed to take this from a WIP to something that is mergeable, unless we continue to hide/flag-guard this new behavior.

**Review by andy-sweet — COMMENTED** [review]
Agreed with @sofroniewn that this is an impressive addition! It also plays nicely with how I imagined people using `Layer.affine` separately from the other components like `Layer.scale`.

I had some extra notes/questions outside of line comments.

- I assume we want the interaction box to update when `Layer.affine` is modified. That doesn't seem to happen yet. Is that as simple as connecting an event to make the interaction box overlay update? Or are there some unknowns/risks?
- The graphical design choices are non-trivial, so I think we should either find a way to get this in without graphically exposing the functionality, or get a little design feedback (e.g. in a community meeting).

File: napari/components/_interaction_box_mouse_bindings.py
Comment (jojoelfe) [reply napari/components/_interaction_box_mouse_bindings.py:139]: >     * I assume we want the interaction box to update when `Layer.affine` is modified. That doesn't seem to happen yet. Is that as simple as connecting an event to make the interaction box overlay update? Or are there some unknowns/risks?

No, should be as simple as that. In my mind, the transform property of the interaction box and the affine property of the layers should be linked, i. e. the controller listens for changes in both and then updates the other one.



>     * The graphical design choices are non-trivial, so I think we should either find a way to get this in without graphically exposing the functionality, or get a little design feedback (e.g. in a community meeting).

Maybe the way to go is to remove the UI elements for now so that the transform mode can only be accessed by setting `.mode` of the layer to `'transform'`. I'll be happy to demo this in a community meeting. Actually there are some other elements I would like to demo that I think are super important to make this an efficient registration system.

**Comment by jojoelfe** [comment]
>     * I assume we want the interaction box to update when `Layer.affine` is modified. That doesn't seem to happen yet. Is that as simple as connecting an event to make the interaction box overlay update? Or are there some unknowns/risks?

No, should be as simple as that. In my mind, the transform property of the interaction box and the affine property of the layers should be linked, i. e. the controller listens for changes in both and then updates the other one.



>     * The graphical design choices are non-trivial, so I think we should either find a way to get this in without graphically exposing the functionality, or get a little design feedback (e.g. in a community meeting).

Maybe the way to go is to remove the UI elements for now so that the transform mode can only be accessed by setting `.mode` of the layer to `'transform'`. I'll be happy to demo this in a community meeting. Actually there are some other elements I would like to demo that I think are super important to make this an efficient registration system.

**Comment by sofroniewn** [comment]
> Maybe the way to go is to remove the UI elements for now so that the transform mode can only be accessed by setting .mode of the layer to 'transform'.

Yeah i think this will be most efficient path

>  I'll be happy to demo this in a community meeting. Actually there are some other elements I would like to demo that I think are super important to make this an efficient registration system.

awesome!!

**Comment by andy-sweet** [comment]
> I think are super important to make this an efficient registration system.

I'd love to hear more about that in the meeting.

Another way to expose this functionality in the GUI (that I've seen in the past) is to have a whole sub-panel to control affine transform parameter values (e.g. a spin box for `translate` for each dimension). But given we have the API and a Python console embedded in napari, I don't think we need something like that and the extra complexity/design that goes with it.

**Comment by brisvag** [comment]
Old related issue: #299.

**Comment by sofroniewn** [comment]
Just had a look through the code, it looks like it's coming along nicely. It seems like you've dropped the qt stuff for now as we agreed.

Changes mostly look good. We can try and get this PR in now for "Image" only but I think we'll want to extend this mode to other layers too - but we can do that in a followup PR.

@jojoelfe are there more changes you'd still like to make here or is it ready for re-review?

**Comment by jojoelfe** [comment]
There are are a few things I am still working on, which maybe could go in a future PR:

- Making sure that updating the affine of the image layer also updates the interaction box: This mostly comes down to making sure that the controller connects to the affine event of the right layer and disconnects at the right time.
- Preserve transform of non-displayed dimensions: For this I think I need a new function in the transform class. Would it be better to open a separate PR or add it to this PR?
- Keyboard shortcut to reset transform

I should be able to get this done over the upcoming holidays, but if you want to review and merge right now I would be fine with that, too.

**Comment by sofroniewn** [comment]
Thanks for the update - all this sounds good @jojoelfe, and some of the make sense in this PR I think. There's no rush, so if you want to do another round that would be great and we can check in then.

**Comment by jojoelfe** [comment]
I think this is ready now for a review. The remaining issue on my end is a bug, where the box behaves like 'Shift' is pressed, which is probably related to #2893 and #2714 .

It probably would good to double-check the behavior in nD. At the moment when the layer is rotated in two dimension, after switching to other dimensions not all slices contain data and the displayed slice changes when the layer is translated. That probably does make some sense, but it is confusing because the displayed slice is no longer determined by the sliders at the bottom.


Also, I am using for now 'Ctrl-Shift-A' to start and stop align/transform mode and 'Ctr-Shift-R' to reset the affine transform of the active layer.

File: napari/_tests/test_interactive_transforms.py
Comment (andy-sweet) [inline napari/_tests/test_interactive_transforms.py]: It's hard to understand the expected value here and in the other test. I think it's all the points in the selection box, including the mid-points and the center.

I would either change things in the code to make this more understandable, or easier define a helper assertion function that some minimal parameterization of that box. For example, you could use the named corners of the box:

```python
def assert_box_has_corners(box, *, top_left, top_right, bottom_right, bottom_left):
   ...
```

File: napari/_tests/test_interactive_transforms.py
Comment (andy-sweet) [inline napari/_tests/test_interactive_transforms.py]: Is the following equivalent?

```suggestion
        viewer.window.qt_viewer.interaction_box_visual.round_marker_node.edge_color.rgba,
```

If so, I would use it, as it uses the public API of vispy and is also more readable.

File: napari/_vispy/overlays/interaction_box.py
Comment (andy-sweet) [inline napari/_vispy/overlays/interaction_box.py]: I think vispy's `ColorArray` can take a numpy array as input, so I think you can do the following instead.

```suggestion
                    face_color = ColorArray(colors[Box.WITH_HANDLE]
```

File: napari/_vispy/overlays/interaction_box.py
Comment (andy-sweet) [inline napari/_vispy/overlays/interaction_box.py]: Same comment as above.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]:
```suggestion
        """Gets called when layer is removed and removes event listener"""

```

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Optional: I think this is equivalent
```suggestion
            getattr(self._viewer.layers.selection.active, 'mode', None) == 'transform'
```
and a little easier to read, mostly because of the long reference to `active`.

Alternatively/Additionally you could pull out a part of that named reference at the beginning (e.g. `selection = self._viewer.layers.selection`), if we trust that `selection` will be a reference to some state and not a copy.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]:
```suggestion
        # This is necessary in case the current layer has fewer dims than the viewer

```

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]:
```suggestion
        # The -0.5 is necessary because the pixel at (0,0) actually extends to (-0.5,0.5) (in case of the image layer)

```

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Optional: consider `getattr` as above.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]:
```suggestion
        """Helper function to connect listener to the transform of active layer and removes previous callbacks"""

```

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]:
```suggestion
            trans._("Reset the affine transform of the active layer"),

```

File: napari/utils/transforms/transforms.py
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
            The transform that will be inserted. Must have as many dimension as len(axes)
```

File: napari/utils/transforms/transforms.py
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
            Axes where the transform will be replaced
```

File: napari/utils/transforms/transforms.py
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]:
```suggestion
            The transform that will be inserted. Must have as many dimension as len(axes)
```

File: napari/utils/transforms/transforms.py
Comment (andy-sweet) [inline napari/utils/transforms/transforms.py]: I'm a little concerned about increasing the size of the public API of the `transforms` module (I think it's already too big!) for one function that is defined for all transform types but is only implemented for one.

As we only use this in one place (i.e. in the interaction box), I suggest one of the following.
1. Define it only for the `Affine` class.
2. Define it as a helper function in the one place it's used.

If you think this function will be used elsewhere (1) is probably better than (2).

File: examples/interaction_box_image.py
Comment (andy-sweet) [inline examples/interaction_box_image.py:11]: This code reduction is great!

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Optional: consider `getattr` as below.

**Review by sofroniewn — APPROVED** [review]
This looks great @jojoelfe. I just read through code, looks ready to go from my point of view. I think next follow up PRs could extend this functionality to all layers (not just image). And then maybe final PR after that would expose more functionality in the GUI/ layer action list.

I've requested some reviews from @andy-sweet @alisterburt @jni maybe some of them can give it a look over and a spin too and then we can merge. Thanks!!

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py:430]: If the active layer is a `Labels` layer and I press ctrl+shift+a, I get an error here because `'transform'` is not a valid mode for the labels layer. Would it be better to use something like `isinstance(active_layer, Image)` here instead?

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Mostly looks great - the implementation is easy enough to follow and the example is so concise! None of my comments are major, but I think there are some minor things to address before we merge this.

One other thing not captured by my comments is that while this responds to mutations of `Layer.affine`, it does not respond to mutations of the properties of `Layers.affine`. I.e. something like the following

```python
viewer.layers[0].affine.scale = [0.5, 2]
```

does not cause the interaction box visual to be updated.

There are no events in the `Affine` instance, so this isn't something you could easily without some significant changes there. So I don't think you should do anything here, but it's worth thinking about in the longer term.

**Review by alisterburt — COMMENTED** [review]
@jojoelfe this is stunning - I love how snappy it feels!

I haven't been over the code in detail - my main comment from playing with the functionality is that I found not being able to zoom when the transform mode was active a little strange.

I understand that this is because pan and zoom are intimately connected. In general, I think we will need to modify the mode/callback management machinery a bit so that it's possible to have multiple callbacks attached in a given mode. Separating pan/zoom or expanding this machinery would definitely be beyond the scope of this PR though and I don't think this should block us moving forward here - this is high value stuff we should aim to get in peoples hands :)

**Comment by sofroniewn** [comment]
> There are no events in the Affine instance, so this isn't something you could easily without some significant changes there. So I don't think you should do anything here, but it's worth thinking about in the longer term.

hmm yeah, we might need to make an issue on that one and followup in a new PR.

> I understand that this is because pan and zoom are intimately connected. In general, I think we will need to modify the mode/callback management machinery a bit so that it's possible to have multiple callbacks attached in a given mode. Separating pan/zoom or expanding this machinery would definitely be beyond the scope of this PR though and I don't think this should block us moving forward here - this is high value stuff we should aim to get in peoples hands :)

yeah, good points. also maybe something for a followup

thanks both for the reviews!!

File: napari/_tests/test_interactive_transforms.py
Comment (jojoelfe) [reply napari/_tests/test_interactive_transforms.py]: I get a `AttributeError: 'Markers' object has no attribute 'edge_color'` when I try that. I tried for a bit to find the face_color in the public API of ther marker visual and could not find it, so I ended up with that code.

Comment (andy-sweet) [reply napari/_tests/test_interactive_transforms.py]: Ah yeah, `edge_color` is a parameter, but not an attribute/property. Sorry!

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [reply napari/components/_interaction_box_mouse_bindings.py]: This has been addressed by using `replace_slice` right? If so, remove the TODO. If not, I think it's fine to leave this and merge without the fix because the new behavior is fairly well hidden behind the keyboard shortcut at the moment.

File: napari/utils/transforms/transforms.py
Comment (jojoelfe) [reply napari/utils/transforms/transforms.py]: I went with (1). I think it will be useful elsewhere, for example for plugins that want to set the affine property in the currently displayed dimensions.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (jojoelfe) [reply napari/components/_interaction_box_mouse_bindings.py:430]: I instead went with enabling the `transform` mode in all layers that have a concept of mode. Happy to revert that and go with the approach you described since it touches code in a lot of places. But it seems quite innocent to me and as a result enables the interaction mode in the shapes, points, and label layer.

**Comment by jojoelfe** [comment]
Thanks for the reviews, I think I addressed everything. @andy-sweet maybe you can check if you agree with everything?

File: napari/_tests/test_interactive_transforms.py
Comment (andy-sweet) [inline napari/_tests/test_interactive_transforms.py]: I like this change because it makes some of the tests below much more readable - hope it makes sense to you too and wasn't too painful to refactor!

However, the box won't be axis-aligned right? So won't you need to check all 4 corners instead of just the top-left and bottom-right?

Also, I would consider making the `corner` arguments keyword only arguments to make it clear what they are when this is being called.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [reply napari/components/_interaction_box_mouse_bindings.py:430]: Sounds good to me!

**Review by andy-sweet — APPROVED** [review]
Thanks for the updates - this looks good to me now, so I'm approving.

I have one remaining question about the box test, but that doesn't block approval.

File: napari/_tests/test_interactive_transforms.py
Comment (jojoelfe) [reply napari/_tests/test_interactive_transforms.py]: Both good points! In these tests the box is axis-aligned, but I renamed the function to make this clear. Also followed your advice to make keyword-only arguments.

**Comment by sofroniewn** [comment]
Ok i think this is all done now, thanks for the work @jojoelfe and reviews @andy-sweet!!

**Pull request merged**
